src/ode_solver/low_storage_runge_kutta_ode_solver.cpp

```diff
@@ -0,0 +1,188 @@
+
+    dealii::LinearAlgebra::distributed::Vector<double> rhs = storage_register_1;
+
+    dealii::LinearAlgebra::distributed::Vector<double> storage_register_4;
+
+    if (this->butcher_tableau->get_b_hat(0) != 0.0){
+        storage_register_4 = storage_register_1;
+    } 
+    
+    double sum_delta = 0;
+
+    //double atol = 0.001;
+    //double rtol = 0.001;
+    double error = 0.0;
+    w = 0.0;
+
+    for (int i = 1; i < n_rk_stages +1; i++ ){
+        // storage_register_2 = storage_register_2 + delta[i-1] * storage_register_1;
+        storage_register_2.add(this->butcher_tableau->get_delta(i-1) , storage_register_1);
+        this->dg->solution = rhs;
+        this->dg->assemble_residual();
+        this->dg->apply_inverse_global_mass_matrix(this->dg->right_hand_side, rhs);
+        // storage_register_1 = gamma[i][0] * storage_register_1 + gamma[i][1] * storage_register_2 + gamma[i][2] * storage_register_3 + beta[i] * dt * rhs;
+        storage_register_1 *= this->butcher_tableau->get_gamma(i, 0);
+        storage_register_1.add(this->butcher_tableau->get_gamma(i, 1), storage_register_2);
+        storage_register_1.add(this->butcher_tableau->get_gamma(i, 2), storage_register_3);
+        rhs *= dt;
+        storage_register_1.add(this->butcher_tableau->get_beta(i), rhs);
+        if (this->butcher_tableau->get_b_hat(i) != 0.0){
+            storage_register_4.add(this->butcher_tableau->get_b_hat(i-1), rhs);
+        }
+        //this->pcout << std::endl;
+        // Check  bhat (i)
+        // rhs = dt * f(S1)
+        rhs = storage_register_1;
+    }
+   // std::abort();
+    // storage_register_2 = (storage_register_2 + delta[m] * storage_register_1 + delta[m+1] * storage_register_3) / sum_delta;
+    if (this->butcher_tableau->get_b_hat(1) == 0){
+        for (int i = 0; i < num_delta; i++){ //change n_rk_stages+2 to num_delta on this line
+        sum_delta = sum_delta + this->butcher_tableau->get_delta(i);
+        }
+        storage_register_2.add(this->butcher_tableau->get_delta(n_rk_stages), storage_register_1);
+        storage_register_2.add(this->butcher_tableau->get_delta(n_rk_stages+1), storage_register_3);
+        storage_register_2 /= sum_delta;
+        // u_hat = s2
+    }
+
+
+    // need to calculate rhs of s1
+    else if (this->butcher_tableau->get_b_hat(n_rk_stages) != 0){
+        this->dg->solution = rhs;
+        this->dg->assemble_residual();
+        this->dg->apply_inverse_global_mass_matrix(this->dg->right_hand_side, rhs);
+        rhs *= dt;
+        storage_register_4.add(this->butcher_tableau->get_b_hat(n_rk_stages), rhs);
+        //this->pcout << " b_hat_fsal " << this->butcher_tableau->get_b_hat(n_rk_stages);
+
+    }
+    this->dg->solution = storage_register_1;
+    
+    double global_size = dealii::Utilities::MPI::sum(storage_register_1.local_size(), this->mpi_communicator);
+    
+    if (this->butcher_tableau->get_b_hat(1) == 0){
```

Comment:
It would be easier to read if you made a `const bool is_3Sstarplus` to control this logic (and above, too). 

Also, I think the computation of `w` would be better to do in `err_time_step`. If you make the storage registers member variables, that should be an easy change!

Reply:
Looks good!